File: internal/ui/theme.go
package ui

import (
	"os"

	"charm.land/lipgloss/v2"
)

// NoColor reports whether color output should be disabled, following the
// NO_COLOR convention (https://no-color.org): the variable is set and non-empty.
func NoColor() bool {
	v, ok := os.LookupEnv("NO_COLOR")
	return ok && v != ""
}

// Theme is the set of styles used across screens. A high-contrast variant avoids
// color and relies on bold/reverse so the UI stays legible without color.
type Theme struct {
	Title    lipgloss.Style
	Subtle   lipgloss.Style
	Normal   lipgloss.Style
	Selected lipgloss.Style
	Accent   lipgloss.Style
	Success  lipgloss.Style
	Error    lipgloss.Style
	Help     lipgloss.Style
	Box      lipgloss.Style
}

// DefaultTheme returns the theme appropriate for the current environment,
// switching to high-contrast when NO_COLOR is set.
func DefaultTheme() Theme {
	return NewTheme(NoColor())
}

// NewTheme builds a theme. When highContrast is true, colors are dropped in
// favor of bold and reverse styling.
func NewTheme(highContrast bool) Theme {
	box := lipgloss.NewStyle().
		Border(lipgloss.RoundedBorder()).
		Padding(1, 3)

	if highContrast {
		return Theme{
			Title:    lipgloss.NewStyle().Bold(true),
			Subtle:   lipgloss.NewStyle().Faint(true),
			Normal:   lipgloss.NewStyle(),
			Selected: lipgloss.NewStyle().Bold(true).Reverse(true),
			Accent:   lipgloss.NewStyle().Bold(true),
			Success:  lipgloss.NewStyle().Bold(true),
			Error:    lipgloss.NewStyle().Bold(true),
			Help:     lipgloss.NewStyle().Faint(true),
			Box:      box,
		}
	}

	var (
		accent  = lipgloss.Color("63")  // indigo
		subtle  = lipgloss.Color("245") // grey
		success = lipgloss.Color("42")  // green
		danger  = lipgloss.Color("203") // red
	)
	return Theme{
		Title:    lipgloss.NewStyle().Foreground(accent).Bold(true),
		Subtle:   lipgloss.NewStyle().Foreground(subtle),
		Normal:   lipgloss.NewStyle(),
		Selected: lipgloss.NewStyle().Foreground(accent).Bold(true),
		Accent:   lipgloss.NewStyle().Foreground(accent),
		Success:  lipgloss.NewStyle().Foreground(success),
		Error:    lipgloss.NewStyle().Foreground(danger),
		Help:     lipgloss.NewStyle().Foreground(subtle),
		Box:      box.BorderForeground(accent),
	}
}

File: term/src/theme.rs
//! Terminal theme: the styles used across screens.
//!
//! Port of the Go `internal/ui` theme. A high-contrast variant avoids color and
//! relies on bold/reverse so the UI stays legible without color (honoring
//! `NO_COLOR`).

use ratatui::style::{Color, Modifier, Style};

/// Reports whether color output should be disabled, following the `NO_COLOR`
/// convention (<https://no-color.org>): the variable is set and non-empty.
pub fn no_color() -> bool {
    std::env::var("NO_COLOR")
        .map(|v| !v.is_empty())
        .unwrap_or(false)
}

/// The set of styles used across screens. Some styles are consumed only by
/// screens still being ported, so the whole palette is retained.
#[allow(dead_code)]
#[derive(Clone, Copy)]
pub struct Theme {
    pub title: Style,
    pub subtle: Style,
    pub normal: Style,
    pub selected: Style,
    pub accent: Style,
    pub success: Style,
    pub error: Style,
    pub help: Style,
    pub border: Style,
}

impl Theme {
    /// The theme appropriate for the current environment, switching to
    /// high-contrast when `NO_COLOR` is set.
    pub fn default_theme() -> Theme {
        Theme::new(no_color())
    }

    /// A theme with no color or text attributes, for deterministic,
    /// escape-free snapshot tests.
    #[allow(dead_code)]
    pub fn plain() -> Theme {
        let p = Style::new();
        Theme {
            title: p,
            subtle: p,
            normal: p,
            selected: p,
            accent: p,
            success: p,
            error: p,
            help: p,
            border: p,
        }
    }

    /// Builds a theme. When `high_contrast` is true, colors are dropped in favor
    /// of bold and reverse styling.
    pub fn new(high_contrast: bool) -> Theme {
        if high_contrast {
            return Theme {
                title: Style::new().add_modifier(Modifier::BOLD),
                subtle: Style::new().add_modifier(Modifier::DIM),
                normal: Style::new(),
                selected: Style::new().add_modifier(Modifier::BOLD | Modifier::REVERSED),
                accent: Style::new().add_modifier(Modifier::BOLD),
                success: Style::new().add_modifier(Modifier::BOLD),
                error: Style::new().add_modifier(Modifier::BOLD),
                help: Style::new().add_modifier(Modifier::DIM),
                border: Style::new(),
            };
        }

        let accent = Color::Indexed(63); // indigo
        let subtle = Color::Indexed(245); // grey
        let success = Color::Indexed(42); // green
        let danger = Color::Indexed(203); // red
        Theme {
            title: Style::new().fg(accent).add_modifier(Modifier::BOLD),
            subtle: Style::new().fg(subtle),
            normal: Style::new(),
            selected: Style::new().fg(accent).add_modifier(Modifier::BOLD),
            accent: Style::new().fg(accent),
            success: Style::new().fg(success),
            error: Style::new().fg(danger),
            help: Style::new().fg(subtle),
            border: Style::new().fg(accent),
        }
    }
}
